[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/cristinaboboc-1975/Seminar-Machine-learning-in-Python/blob/main/Tema_Laborator_1_ML.ipynb)

# 📝 Tema de Casă / Exercițiu Individual — Laborator 1
**Curs:** Machine Learning în Python  
**Subiect:** Manipularea datelor cu Python, NumPy și Pandas  
**Timp estimat:** 30–45 minute | **Punctaj total:** 10 puncte

---

### 🎯 Contextul Problemei
Sunteți Data Scientist într-un proiect educațional. Ați primit un set de date care conține rezultatele mai multor studenți la 3 evaluări intermediare (Test 1, Test 2, Proiect), precum și numărul de ore studiate săptămânal.

Obiectivul vostru este să analizați și să pregătiți aceste date pentru un viitor model de Machine Learning.

### 🔹 Cerința 1: Funcții Python & Logică (2 puncte)
Scrieți o funcție Python numită `calculeaza_nota_finala(t1, t2, proiect)` care primește 3 note (numere reale între 1 și 10) și calculează media ponderată după următoarea regulă:
- **Test 1:** pondere de 30% (`0.30`)
- **Test 2:** pondere de 30% (`0.30`)
- **Proiect:** pondere de 40% (`0.40`)

*Cerință:* Funcția trebuie să returneze media ponderată rotunjită la două zecimale. Testați funcția pentru notele: `Test 1 = 8.5`, `Test 2 = 7.0`, `Proiect = 9.5`.

In [1]:
# Rezolvare Cerința 1
def calculeaza_nota_finala(t1, t2, proiect):
    media = t1 * 0.30 + t2 * 0.30 + proiect * 0.40
    return round(media, 2)

# Testare funcție
print(calculeaza_nota_finala(8.5, 7.0, 9.5))

8.45


### 🔹 Cerința 2: Operare cu Vectori în NumPy (3 puncte)
Avem următorul vector NumPy care reprezintă orele de studiu săptămânale ale 8 studenți:
```python
import numpy as np
ore_studiu = np.array([12, 5, 18, 8, 15, 3, 22, 10])
```

1. Calculați **media**, **deviația standard**, valoarea **minimă** și valoarea **maximă** a orelor de studiu folosind funcții din `numpy`.
2. Creați un filtru boolean care să identifice studenții care au studiat **mai mult de 10 ore** pe săptămână.
3. Afișați doar orele de studiu ale studenților care îndeplinesc această condiție.

In [9]:
# Rezolvare Cerința 2
import numpy as np

ore_studiu = np.array([12, 5, 18, 8, 15, 3, 22, 10])

# 1. Statistici descriptive
# Media aritmetica
media = np.mean(ore_studiu)

# Deviația standard
deviatia_std = np.std(ore_studiu)

# Valoarea minima si maxima
minim = np.min(ore_studiu)
maxim = np.max(ore_studiu)

# Afisaree
print("Media este:", media, "ore")
print("Deviația standard este:", round(deviatia_std, 2))
print("Valoarea minima este:", minim, "ore")
print("Valoarea maxima este:", maxim, "ore")

# 2. Filtru boolean
filtru = ore_studiu > 10
print("Filtru:", filtru)

# 3. Orele studenților care au studiat mai mult de 10 ore
print("Orele studenților care au studiat mai mult de 10 ore:", ore_studiu[filtru])

Media este: 11.625 ore
Deviația standard este: 6.06
Valoarea minima este: 3 ore
Valoarea maxima este: 22 ore
Filtru: [ True False  True False  True False  True False]
Orele studenților care au studiat mai mult de 10 ore: [12 18 15 22]


### 🔹 Cerința 3: Creare și Manipulare Tabel Pandas (4 puncte)
Folosind dicționarul oferit, creați un `DataFrame` Pandas numit `df_studenti` și realizați următoarele prelucrări:

1. Adăugați o coloană nouă în tabel, numită `Nota_Finala`, calculată aplicând funcția `calculeaza_nota_finala` creată la **Cerința 1** peste coloanele `Test_1`, `Test_2` și `Proiect`.
2. Adăugați o coloană booleană `Admis` care conține valoarea `True` dacă `Nota_Finala >= 6.0` și `False` în caz contrar.
3. Filtrați tabelul și afișați doar studenții **Admiși** care au învățat **peste 10 ore** (`Ore_Studiu > 10`).

In [15]:
# Rezolvare Cerința 3
import pandas as pd

# copiem functia de am scris-o anterior mai sus la cealalta cerinta
def calculeaza_nota_finala(t1, t2, proiect):
    media = t1 * 0.30 + t2 * 0.30 + proiect * 0.40
    return round(media, 2)

date_studenti = {
    'Nume': ['Ana', 'Bogdan', 'Cristian', 'Elena', 'Florin', 'Gabriela', 'Horia', 'Ioana'],
    'Test_1': [8.5, 5.0, 9.0, 6.5, 4.0, 9.5, 7.0, 8.0],
    'Test_2': [7.0, 6.0, 8.5, 5.5, 4.5, 10.0, 6.5, 8.5],
    'Proiect': [9.5, 7.0, 9.5, 8.0, 5.0, 9.5, 7.5, 9.0],
    'Ore_Studiu': [12, 5, 18, 8, 15, 3, 22, 10]
}

# cream tabel
df_studenti = pd.DataFrame(date_studenti)

# 1. Calcul Nota_Finala

df_studenti['Nota_Finala'] = df_studenti.apply(
    lambda rand: calculeaza_nota_finala(rand['Test_1'], rand['Test_2'], rand['Proiect']),
    axis=1
)


# 2. Adăugare coloană Admis
df_studenti['Admis'] = df_studenti['Nota_Finala'] >= 6.0

# 3. Filtrare
df_filtrat = df_studenti[(df_studenti['Admis']) & (df_studenti['Ore_Studiu'] > 10)]

#afisare
print(df_studenti)
print(' ')
print(df_filtrat)

       Nume  Test_1  Test_2  Proiect  Ore_Studiu  Nota_Finala  Admis
0       Ana     8.5     7.0      9.5          12         8.45   True
1    Bogdan     5.0     6.0      7.0           5         6.10   True
2  Cristian     9.0     8.5      9.5          18         9.05   True
3     Elena     6.5     5.5      8.0           8         6.80   True
4    Florin     4.0     4.5      5.0          15         4.55  False
5  Gabriela     9.5    10.0      9.5           3         9.65   True
6     Horia     7.0     6.5      7.5          22         7.05   True
7     Ioana     8.0     8.5      9.0          10         8.55   True
 
       Nume  Test_1  Test_2  Proiect  Ore_Studiu  Nota_Finala  Admis
0       Ana     8.5     7.0      9.5          12         8.45   True
2  Cristian     9.0     8.5      9.5          18         9.05   True
6     Horia     7.0     6.5      7.5          22         7.05   True


### 🔹 Cerința 4: Întrebare de Reflexie (1 punct)
Privind tabelul final obținut:  
*Existe vreun student care a obținut o notă mare (ex. peste 9.0), dar a învățat un număr foarte mic de ore (ex. sub 5 ore)? Cum ați explica acest caz dacă ar fi vorba despre un set de date din lumea reală și ce impact ar avea un astfel de caz asupra unui model de Machine Learning?*

In [16]:
print(df_studenti[(df_studenti['Nota_Finala'] > 9.0) & (df_studenti['Ore_Studiu'] < 5)])

       Nume  Test_1  Test_2  Proiect  Ore_Studiu  Nota_Finala  Admis
5  Gabriela     9.5    10.0      9.5           3         9.65   True


**Răspunsul tău pentru Cerința 4:**  
*Da, studenta Gabriela are 9.65 si a studiat doar 3 ore. In lumea reala ar putea insemna ca studenta are un talent sau o pregătire anterioară foarte bună și îi trebuie mai puține ore pentru aceleași rezultate, sau poate fi si eroare in introducerea datelor. Pot exista si factori externi spre exemplu, experienta in domeniu (poate lucreaza in acest domeniu si are mai multe cunostinte).*

*Ca impact asupra unui model Machine Learning, acesta este un outlier si poate distorsiona modelul creat. De exemplu, într-o regresie liniară, un punct cu ore puține și notă mare „trage" dreapta și slăbește relația dintre ore și notă, ceea ce duce la predicții mai slabe pentru restul studenților. Dar depinde si de marimea setului de date. Daca e mic, un sigur caz de genul cum e in exmplu, cantareste mult.*